[FEATURE][CERTIF] Permettre l'inscription à Pix Certif depuis une invitation (PIX- 6209)

annemarie35 opened the pull request:
## :jack_o_lantern: Problème

Dans le cadre de l'epix, nous voulons permettre aux agents Pix d'envoyer par e-mail des invitations à rejoindre un centre de certification depuis Pix Admin. Il se peut que le destinataire de l'e-mail n'ait pas déjà un compte Pix, auquel cas il doit pouvoir le créer depuis la double-mire.

## :bat: Proposition

Permettre la création d'un compte depuis la double mire lorsque l'utilisateur suit le lien d'une invitation.

## :spider_web: Remarques
- https://github.com/1024pix/pix/pull/2278
- https://github.com/1024pix/pix/pull/4080
- Ce ticket provient d'un split de la PR https://github.com/1024pix/pix/pull/5129 afin de créer la route api pour permettre de traiter 2 tickets front qui en dépendent en parallèle.

## :ghost: Pour tester

#### Invitation qui n'existe pas

- Allez sur le lien https://certif-pr5157.review.pix.fr/rejoindre?invitationId=111&code=AAA
- Constatez que vous arrivez sur la page de connexion de l'application Pix Certif

#### Invitation existante et en attente

- Allez sur le lien https://certif-pr5157.review.pix.fr/rejoindre?invitationId=101259&code=ABCDEF123
- Constatez que vous arrivez sur la double mire
- Fournir toutes les informations nécessaire à l'inscription (Prénom, Nom, Adresse e-mail, Mot de passe)
- Acceptez les CGU (checkbox)
- Cliquez sur le bouton `Je m'inscris`
- Constatez que vous arrivez sur la page des CGU (`/cgu`)
- Cliquez sur le bouton `J'accepte les conditions d'utilisation`
- Constatez que vous arrivez sur la liste des sessions de certification 🎉

#### Invitation existante, en attente et utilisation d'un e-mail existant

- Allez sur le lien https://certif-pr5157.review.pix.fr/rejoindre?invitationId=101260&code=ABCDEF123
- Constatez que vous arrivez sur la double mire
- Fournir toutes les informations nécessaire à l'inscription (Prénom, Nom, Adresse e-mail, Mot de passe) avec un e-mail existant en base de données
- Acceptez les CGU (checkbox)
- Cliquez sur le bouton `Je m'inscris`
- Constatez que vous êtes toujours sur la double mire
- Constatez qu'un message d'erreur est affiché `Cette adresse e-mail est déjà enregistrée, connectez-vous.`

#### Invitation existante et déjà acceptée

- Allez sur le lien https://certif-pr5157.review.pix.fr/rejoindre?invitationId=101261&code=ABCDEF123
- Constatez que vous arrivez sur la page de connexion de l'application Pix Certif

#### Invitation existante et annulée

- Allez sur le lien https://certif-pr5157.review.pix.fr/rejoindre?invitationId=101262&code=ABCDEF123
- Constatez que vous arrivez sur la page de connexion de l'application Pix Certif

AndreiaPena committed:
feat(certif): create isFormValid method and add into register method

AndreiaPena committed:
feat(certif): create pix user account

AndreiaPena committed:
feat(certif): create certification-center-invitation-response adapter

AndreiaPena committed:
feat(certif): create membership and validate invitation

AndreiaPena committed:
feat(certif): authenticate user

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr5157.review.pix.fr
- App (.org): https://app-pr5157.review.pix.org
- Orga: https://orga-pr5157.review.pix.fr
- Certif: https://certif-pr5157.review.pix.fr
- Admin: https://admin-pr5157.review.pix.fr
- API: https://api-pr5157.review.pix.fr/api/

 Please check it out!

igorissen committed:
feat(certif): pass certification parameters to children components

igorissen committed:
sr(api): send 204 status code instead of 200

There is no need to send User data because we are using another model (CertificationPointOfContact) in Ember application to make the request

Co-authored-by: Anne-G <ang.schall@gmail.com>

igorissen committed:
test(certif): add missing acceptance test

Co-authored-by: Anne-G <ang.schall@gmail.com>

igorissen committed:
feat(certif): display a notification

when an error occurs

Co-authored-by: Anne-G <ang.schall@gmail.com>

igorissen committed:
sr(certif): move notifications container in application template

Co-authored-by: Anne-G <ang.schall@gmail.com>

annemarie35 commented on certif/tests/unit/components/auth/register-form_test.js:
```diff
@@ -0,0 +1,102 @@
+        await component.register(eventStub);
+
+        // then
+        assert.ok(component.store.createRecord.notCalled);
```
suggestion : utiliser le calledWith pour vérifier qu'on appelle avec les bons paramètres

annemarie35 commented on certif/tests/acceptance/routes/join_test.js:
```diff
@@ -0,0 +1,35 @@
+      await clickByName(`J’accepte les conditions d’utilisation`);
+
+      // then
+      assert.strictEqual(currentURL(), '/sessions/liste');
```
suggestion : vérifier qu'il est bien sur le bon centre de certif

igorissen committed:
refactor(certif): set overflow-y to auto  instead of scroll

annemarie35 commented on certif/app/components/auth/register-form.js:
```diff
@@ -33,10 +35,38 @@ export default class RegisterForm extends Component {
+        this.errorMessage = this.intl.t('pages.login-or-register.register-form.errors.default');
+      }
+
+      await userRecord?.deleteRecord();
```
Remarque :

annemarie35 commented on certif/app/components/auth/register-form.js:
```diff
@@ -53,6 +54,18 @@ export default class RegisterForm extends Component {
+        adapterOptions: { certificationCenterInvitationId: this.args.certificationCenterInvitationId },
+      });
+
       this.password = null;
```
Chipotage : ajouter un test sur le this.password= null; 

ex 
`assert.strictEqual(component.password, null);`

annemarie35 replied:
Si ce bout de code est important ^^

igorissen replied:
En soit ce bout de code pourrait être retiré parce qu'il n'a pas de sens. Vu que l'on connecte l'utilisateur donc il se redirigé vers une nouvelle page.

annemarie35 commented on certif/app/controllers/terms-of-service.js:
```diff
@@ -4,13 +4,18 @@ import { inject as service } from '@ember/service';
+      await this.currentUser.certificationPointOfContact.save({
+        adapterOptions: { acceptPixCertifTermsOfService: true },
+      });
+      this.currentUser.certificationPointOfContact.pixCertifTermsOfServiceAccepted = true;
```
Remarque, j'ai pas eu besoin de faire ça pour la connexion.
l'apple à http://localhost:4203/api/users/104439/pix-certif-terms-of-service-acceptance met les cgu à true
je suis pas allée voir la page cgu mais elle fonctionne déjà pour les utilisateurs pix certif (si un utilisateur se loggue et que ses cgu certif sont à false, il passe par cette page).
je pense que tu peux supprimer ce bout de code 
sauf si tu as vu un cas ?

igorissen replied:
Cet appel retourne une 204. Le modèle `CertificationPointOfContact` ne sera pas modifié au niveau du front. Je pense que ce code est nécessaire.

Pour ce fichier, nous avons juste englobé le code dans un `try-catch` pour pouvoir afficher une notification en cas d'erreur.

annemarie35 replied:
Ah mais oui, j'avais pas bien vu, je retire ce que j'ait dit

annemarie35 reviewed (commented):
Ceci est un commentaire

annemarie35 reviewed (commented):
Approved !
